stablehlo/dialect/StablehloBytecode.cpp

```diff
@@ -0,0 +1,695 @@
+      DialectBytecodeReader &reader) const;
+  TransposeAttr readTransposeAttr(DialectBytecodeReader &reader) const;
+  TypeExtensionsAttr readTypeExtensionsAttr(
+      DialectBytecodeReader &reader) const;
+
+  // TO ADD ATTRIBUTE: Include a write method for each attribute in StableHLO
+  // Ex: void write(SomeAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ArgResultAliasAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ChannelHandleAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ComparisonDirectionAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ComparisonTypeAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ConvDimensionNumbersAttr attr,
+             DialectBytecodeWriter &writer) const;
+  void write(DotDimensionNumbersAttr attr, DialectBytecodeWriter &writer) const;
+  void write(FftTypeAttr attr, DialectBytecodeWriter &writer) const;
+  void write(GatherDimensionNumbersAttr attr,
+             DialectBytecodeWriter &writer) const;
+  void write(PrecisionAttr attr, DialectBytecodeWriter &writer) const;
+  void write(RngAlgorithmAttr attr, DialectBytecodeWriter &writer) const;
+  void write(RngDistributionAttr attr, DialectBytecodeWriter &writer) const;
+  void write(ScatterDimensionNumbersAttr attr,
+             DialectBytecodeWriter &writer) const;
+  void write(TransposeAttr attr, DialectBytecodeWriter &writer) const;
+  void write(TypeExtensionsAttr attr, DialectBytecodeWriter &writer) const;
+
+  //===--------------------------------------------------------------------===//
+  // Types
+
+  // These methods are invoked by superclass when a type from stablehlo dialect
+  // is encountered.
+  Type readType(DialectBytecodeReader &reader) const override;
+  LogicalResult writeType(Type type,
+                          DialectBytecodeWriter &writer) const override;
+
+  // TO ADD TYPE: Include a read method for each type in StableHLO
+  // Ex: SomeType readSomeType(DialectBytecodeReader &reader) const;
+  TokenType readTokenType(DialectBytecodeReader &reader) const;
+
+  // TO ADD TYPE: Include a write method for each type in StableHLO
+  // Ex: void write(SomeType attr, DialectBytecodeWriter &writer) const;
+  void write(TokenType type, DialectBytecodeWriter &writer) const;
+
+ private:
+  //===--------------------------------------------------------------------===//
+  // Helper methods
+
+  // Enum reader and writer. Many attrs have a single enum type to serialize.
+  // Use the attributes underlying type to get the numeric value.
+  // Note this may cause issues if enums use an int64_t and have a large value.
+  // All enums in StableHLO currently use int32_t.
+  template <typename EnumType, typename EnumTypeAttr, typename SymbolizeFn>
+  EnumTypeAttr readEnumAttribute(DialectBytecodeReader &reader,
+                                 SymbolizeFn symbolizeFn) const {
+    uint64_t code;
+    if (failed(reader.readVarInt(code))) return EnumTypeAttr();
```

Comment:
Is this suggesting to use `return {};` instead of the empty `return SomeAttr();`? If so this decision was based on how BuiltinDialect implemented this:
https://github.com/llvm/llvm-project/blob/main/mlir/lib/IR/BuiltinDialectBytecode.cpp#L443

Reply:
Interesting! Nevermind then I guess.